addons/stock/models/product_strategy.py

```diff
@@ -12,47 +12,36 @@ class RemovalStrategy(models.Model):
+class PutAwayRule(models.Model):
     _name = 'product.putaway'
-    _description = 'Put Away Strategy'
-
-    name = fields.Char('Name', required=True)
-
-    fixed_location_ids = fields.One2many(
-        'stock.fixed.putaway.strat', 'putaway_id',
-        'Fixed Locations Per Product Category', domain=[('category_id', '!=', False)], copy=True)
-    product_location_ids = fields.One2many(
-        'stock.fixed.putaway.strat', 'putaway_id',
-        'Fixed Locations Per Product', domain=[('product_id', '!=', False)], copy=True)
-
-    def putaway_apply(self, product):
-        put_away = self._get_putaway_rule(product)
-        if put_away:
-            return put_away.fixed_location_id
-        return self.env['stock.location']
-
-    def _get_putaway_rule(self, product):
-        if self.product_location_ids:
-            put_away = self.product_location_ids.filtered(lambda x: x.product_id == product)
-            if put_away:
-                return put_away[0]
-        if self.fixed_location_ids:
-            categ = product.categ_id
-            while categ:
-                put_away = self.fixed_location_ids.filtered(lambda x: x.category_id == categ)
-                if put_away:
-                    return put_away[0]
-                categ = categ.parent_id
-        return self.env['stock.location']
-
-
-class FixedPutAwayStrategy(models.Model):
-    _name = 'stock.fixed.putaway.strat'
     _order = 'sequence'
-    _description = 'Fixed Putaway Strategy on Location'
+    _description = 'Putaway Rule'
 
     product_id = fields.Many2one('product.product', 'Product')
-    putaway_id = fields.Many2one('product.putaway', 'Put Away Method', required=True)
     category_id = fields.Many2one('product.category', 'Product Category')
-    fixed_location_id = fields.Many2one('stock.location', 'Location', required=True)
+    location_in_id = fields.Many2one('stock.location', 'When product arrives in', required=True)
+    location_out_id = fields.Many2one('stock.location', 'Store to', required=True)
     sequence = fields.Integer('Priority', help="Give to the more specialized category, a higher priority to have them in top of the list.")
+
+    product_tmpl_id = fields.Many2one('product.template', compute='_compute_product_template_id',
+        string="Product Template", store=True)
+
+    @api.multi
+    @api.depends('product_id')
+    def _compute_product_template_id(self):
+        for record in self:
+            if record.product_id:
+                record.product_tmpl_id = record.product_id.product_tmpl_id
+
+    def putaway_for_product(self, product):
+        if product == self.product_id:
+            return self.location_out_id
+        return self.env['stock.location']
+
+    def putaway_for_category(self, product):
+        category = product.categ_id
+        while category:
+            if category == self.category_id:
+                return self.location_out_id
+            category = category.parent_id
```

Comment:
it would be nice to not have any method here, try to move everything to get_putaway_strategy in stock location